### ISSUE TITLE
Allow icon customization without webpack-specific plugins

### ISSUE THREAD

**Issue by filipsobol** [ckeditor/ckeditor5#16546]
Currently, the only way to customize icons is to use webpack-specific `NormalModuleReplacementPlugin`. However, this doesn't work with the [new installation methods](https://github.com/ckeditor/ckeditor5/issues/15502), which are bundler-agnostic.

We have had some discussions about how to allow replacing icons in NIM, and the first step would be to move all icons to a new package that only re-exports icons individually (not grouped into a single JavaScript object). This would allow icons to be tree-shaken by the bundlers when not used.

The next step is not clear yet, as there are two possible approaches:

1.  Icons should behave like new translations and be passed to the editor configuration. This way, most people can pass the default icons, but those who want customization can pass a custom set of icons.
2.  Plugins should import icons from the new package, and the developers who want to change them can use the `override` option available in [npm](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides) and [pnpm](https://pnpm.io/package_json#pnpmoverrides) or `resolutions` available in [yarn](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/) to override the icons package with their own custom package.

The first approach provides a better developer experience at the cost of introducing breaking changes for all developers using CKEditor5.  
The second approach introduces a smaller breaking change that doesn't affect the majority of the projects, but at the cost of the worst developer experience for those who customize icons.

**Comment by trevordevore** [comment]
I vote for 1. Worst developer experience will remain forever. The one-time code change is a... one-time change.

**Comment by sunesimonsen** [comment]
I think option one sounds best.

Here is an idea - what if CKEditor had default icons (like it does today), but you can override them with an icon map you pass into the configuration. Wouldn't that maintain backward compatibility for most usages, while giving the power to override any icon?

**Comment by filipsobol** [comment]
> Here is an idea - what if CKEditor had default icons (like it does today), but you can override them with an icon map you pass into the configuration. Wouldn't that maintain backward compatibility for most usages, while giving the power to override any icon?

Yes, this would maintain backwards compatibility and allow overriding icons. However, in this scenario, the final bundle would include both icon sets, which we want to avoid.

**Comment by pomek** [comment]
Done in https://github.com/ckeditor/ckeditor5/pull/17750.

**Comment by Siddhant-Sentieo** [comment]
Is there a workaround to make it work using webpack NormalModuleReplacementPlugin somehow before the fix gets released ?

**Comment by filipsobol** [comment]
> Is there a workaround to make it work using webpack NormalModuleReplacementPlugin somehow before the fix gets released ?

Sorry, we need more context. Are you talking about some specific issue? If so, can you link it?

**Comment by Siddhant-Sentieo** [comment]
> > Is there a workaround to make it work using webpack NormalModuleReplacementPlugin somehow before the fix gets released ?
> 
> Sorry, we need more context. Are you talking about some specific issue? If so, can you link to it?

I was talking about this issue: https://github.com/ckeditor/ckeditor5/issues/16546, where Icon Customisation is not possible in the new installation methods.

Just to give you context, In our application we have created a separate package/micro-app for serving CKEditor build to other components. In this package we are using webpack as a module bundler.
But after upgrading the editor to the latest version, NormalModuleReplacementPlugin is not replacing the toolbar icons with our custom icons. Is there any temporary workaround to make this work ?

**Comment by filipsobol** [comment]
Currently, there's no workaround for this in new installation methods (see [Feature comparison](https://ckeditor.com/docs/ckeditor5/latest/updating/nim-migration/migration-to-new-installation-methods.html#feature-comparison)).

We made necessary changes to reintroduce support for customizing icons, and it will be available by the end of Q1 2025.
